bambulabs_api/filament_info.py

```
@@ -26,8 +26,37 @@ class AMSFilamentSettings:
     nozzle_temp_max: int
     tray_type: str
 
-
-class Filament(AMSFilamentSettings, Enum):
+class Filament(AMSFilamentSettings):
+    """Changes the behavior of the Filament library class to allow unknown/new filaments (tray_info_idx codes) to be set. 
+    Does not validate manually-set the filament codes against a known list, but printer should just safely/silently ignore it if it's not real"""
+
+    def __init__(self, filament: str | AMSFilamentSettings):
+        
+        # Support existing lookup by standard filament names known to the library
+        if type(filament) is str:
+            name = filament 
+            print(name)
+            
+            # This will throw an exception if the name can't be mapped to a known setting, just like before
+            knownF = FilamentEnum(name)
+            
+            super(Filament, self).__init__(
+                knownF.tray_info_idx,
+                int(knownF.nozzle_temp_min),
+                int(knownF.nozzle_temp_max),
+                knownF.tray_type,
+                )
+        
+        # Support creating your own arbitrary AMSFilamentSettings so newer/custom codes can be used
+        else:
+            super(Filament, self).__init__(
+                filament.tray_info_idx,
```

Not sure how the printers will handle a custom index
@ohmdelta any idea?

---

For what it's worth my A1 just silently ignored the request if it didn't recognize the index.

---

I think the custom index is fine.

The Filament enum class (that you've renamed to FilamentEnum here) probably shouldn't be used as it has been in the library. We'll probably just drop the references to Filament enum class.

@Aptimex @mchrisgm - #130

---

Cool, yeah that's looks like a cleaner solution.